feat: use jsipfs as fallback instead of public gateways

fsdiogo opened the pull request:
Closes #43.

victorb commented on src/bundles/files.js:
```diff
@@ -243,15 +296,17 @@ export default {
       }
 
       for (const file of ipfsFiles) {
+        const fileId = shortid.generate()
```
Why not use the hash as the ID instead of a shortid?

fsdiogo replied:
I could, but I prefer to keep the logic consistent with the add files one (when adding files I'm using shortId because I only get the hash after the file is added).

fsdiogo commented:
When a file larger than 1GB is found, a warning is shown: (in the image I mocked it to be for files larger than 2MB)

<img width="1274" alt="warning" src="https://user-images.githubusercontent.com/33324750/48718721-d791f200-ec13-11e8-8d23-6edb6a252e7a.png">

fsdiogo commented:
Fixed the broken link and updated the copy, thanks @lidel!

About the open problem: I agree that it needs to be addressed ASAP, but always and not only when using an embedded `js-ipfs` node. Even if the user is using a local node, if he adds a file, gets the share link and then quits its daemon, the file is "lost". That must be explicit somewhere, and I think it should be somewhere in the copy of "how the share app works".

olizilla commented:
👀

olizilla commented:
> FUN: I noticed ipfs-redux-bundle by default also tries to connect to local API port directly – not sure if it is a bug... it could be a feature ;)

@lidel that was intentional, is it a problem?

lidel commented:
@olizilla not a problem, I think it makes sense for Share app to do such opportunistic lookup.
However the [error indicated by the first arrow here](https://user-images.githubusercontent.com/157609/48777671-89dcbe80-ecd3-11e8-8734-c9d22c52810b.png) suggests we are using multiaddr without converting it to URL first somewhere (it then retries with real URL).

olizilla commented on public/locales/pt/translation.json:
```diff
@@ -32,6 +33,6 @@
   "downloadFiles": {
     "downloadAll": "Descarregar todos",
-    "downloading": "A arquivar..."
+    "downloading": "A descarregar..."
```
@fsdiogo just to check, did these pt language changes get pulled from transifex, or did you edit them in the file manually? Do they need to be pushed to transifex?

fsdiogo replied:
This project isn't on transifex yet; I was waiting to get some sort of "final copy" first because it's possible that many of these tags change.

olizilla commented on src/bundles/files.js:
```diff
@@ -262,10 +340,51 @@ export default {
+              dispatch({ type: 'FILES_DOWNLOAD_PROGRESS', payload: { id: id, progress: progress } })
+              chunks.push(chunk)
+            })
+            .on('end', () => {
```
I like that you've done this without adding additional imports, but we could potentially use the Buffer instance on ipfs to concat the chunks in fewer lines (on our side at least) with

```suggestion
            .on('end', () => {
              const buffer = ipfs.types.Buffer.concat(chunks)
              dispatch({ type: 'FILES_DOWNLOAD_FINISHED', payload: { id: id } })
              resolve(buffer)
            })
```

https://github.com/ipfs/js-ipfs-api#domain-data-types
https://github.com/ipfs/js-ipfs#domain-data-types

though, as a type that, I wonder if `window.ipfs` has those...

olizilla replied:
Not yet https://github.com/ipfs-shipyard/ipfs-postmsg-proxy/issues/20

fsdiogo replied:
I like that. When #20 is done I'll refactor this.

olizilla commented on src/components/download-files/DownloadFiles.js:
```diff
@@ -12,40 +11,38 @@ import 'react-circular-progressbar/dist/styles.css'
+
+    for (const file of Object.values(files)) {
+      const fileContent = await doDownloadFile(file.id, file.hash)
+      downloadFile(fileContent, file.name)
```
`downloadFile` and `doDownloadFile` do very different things, but have almost the same name. Perhaps

- `doDownloadFile` -> `doGetFromIPFS`
- `downloadFile` -> `triggerSaveFile`

or anything more specific.

olizilla replied:
Changing either one of them would be enough to make it easier to understand the difference.

fsdiogo replied:
SGTM. I'll update this in #47 to avoid merge conflicts.

olizilla commented on src/components/file/File.js:
```diff
@@ -27,47 +28,55 @@ export class File extends React.Component {
-    progress: null
+  handleDownloadClick = async () => {
+    const { id, hash, name, doDownloadFile } = this.props
+    const file = await doDownloadFile(id, hash)
```
Thinking out loud, and for another PR, perhaps we should start fetching the files from IPFS eagerly, before the user clicks on them.

fsdiogo replied:
That would be a nice optimisation 🤔

I'll keep this in mind and maybe open an issue about it later!

olizilla commented:
@lidel thanks! now i see it. That is gross. In part it's a bug from https://github.com/tableflip/uri-to-multiaddr - I assumed that dnsaddr was a generic protocol for "could be dns4 or dns6", but it actually means, check this domain for a dnslink! The only reference I can find to back that up is https://github.com/multiformats/go-multiaddr-dns/issues/7#issuecomment-375535931 which needs more documenting.

The next problem is there seems to be no way to convert a uri to a multiaddr correctly without being able to do dns lookups or know in advance if a domain will resolve to an. ipv4 or ipv6 address.

lidel commented:
@olizilla 

I was going to suggest that ip4/ip6 hint (based on presence of a specific DNS record type: `A` vs `AAA`) could be added to the output of DNSLink lookup endpoint at `/api/v0/dns/` (the other one being TTL), but we experience the error while resolving API address :man_facepalming:  and asking https://ipfs.io/api/v0/dns/ is bit ugly.

So.. should we switch default API multiaddr to `/ip4/` ?
